### DISCUSSION TITLE
Pan without ctrl/meta key

### DISCUSSION THREAD

**Discussion by arwtyxouymz** [samuelarbibe/dnd-timeline#55] [Q&A] [answered]
Hi
This library is what I wanted! Thank you for your great work!

I have a question about pan strategy.
I want to scroll timeline vertically and horizontally by dragging without ctrl/meta key.
Is it possible to achieve it?

This is the current behavior.
Horizontal scroll is working, but vertical scroll is not.
It scales timeline range and select sidebar text.

https://github.com/user-attachments/assets/3e6c6c58-a2de-4cd2-b71c-39acca713810

This is my current `UsePanStrategy`. I referred to your default `useDragStrategy`.
I think other resource (probably `useItem` hook) event handler is triggered at the same time.
I appreciate your help and  I'm also happy to help contribute to this request if you'd like any assistance.
```ts
import { useLayoutEffect, useRef } from 'react';
import type { PanEndEvent, UsePanStrategy } from 'dnd-timeline';

export const useDragStrategy: UsePanStrategy = (timelineRef, onPanEnd) => {
  const lastDragPos = useRef<{ x: number; y: number } | null>(null);

  useLayoutEffect(() => {
    const element = timelineRef.current;
    if (!element) return;

    const pointerdownHandler = (event: MouseEvent) => {
      lastDragPos.current = { x: event.clientX, y: event.clientY };
    };

    const pointerupHandler = () => {
      lastDragPos.current = null;
    };

    const pointermoveHandler = (event: MouseEvent) => {
      if (!lastDragPos.current) return;

      const deltaX = event.clientX - lastDragPos.current.x;
      const deltaY = event.clientY - lastDragPos.current.y;

      lastDragPos.current = { x: event.clientX, y: event.clientY };

      const panEndEvent: PanEndEvent = {
        deltaX: -deltaX,
        deltaY: -deltaY,
      };

      onPanEnd(panEndEvent);
    };

    element.addEventListener('pointerup', pointerupHandler);
    element.addEventListener('pointerdown', pointerdownHandler);
    element.addEventListener('pointermove', pointermoveHandler);
    element.addEventListener('pointerleave', pointerupHandler);

    return () => {
      element.removeEventListener('pointerup', pointerupHandler);
      element.removeEventListener('pointerdown', pointerdownHandler);
      element.removeEventListener('pointermove', pointermoveHandler);
      element.removeEventListener('pointerleave', pointerupHandler);
    };
  }, [onPanEnd, timelineRef]);
};
```

**Comment by samuelarbibe** [comment, marked as answer]
Hey!  
Vertical and Horizontal scrolling behave differently because the vertical scrolling is a pure CSS scroll, and the horizontal scrolling is a controlled scroll, that controls a `range` state.  
  
Therefore, the `range` state has to be controlled by a custom hook, while the vertical works without any intervention.  
  
There are ways to control CSS controlled scroll, Here is an example:  
```tsx

export const useDragStrategy: UsePanStrategy = (timelineRef, onPanEnd) => {
	const lastDragX = useRef<number | null>(null);
	const lastDragY = useRef<number | null>(null);

	useLayoutEffect(() => {
		const element = timelineRef.current;
		if (!element) return;

		const pointerdownHandler = (event: MouseEvent) => {
			if (!event.ctrlKey && !event.metaKey) return;

			lastDragX.current = event.clientX;
			lastDragY.current = event.clientY;
		};

		const pointerupHandler = () => {
			lastDragX.current = null;
			lastDragY.current = null;
		};

		const pointermoveHandler = (event: MouseEvent) => {
			if (!lastDragX.current || !lastDragY.current) return;

			const deltaX = event.clientX - lastDragX.current;
			const deltaY = event.clientY - lastDragY.current;

			lastDragX.current = event.clientX;
			lastDragY.current = event.clientY;

			const panEndEvent: PanEndEvent = {
				deltaX: -deltaX,
				deltaY: 0,
			};

			timelineRef.current?.scroll({
				top: timelineRef.current.scrollTop - deltaY,
				behavior: "instant",
			});

			onPanEnd(panEndEvent);
		};

		element.addEventListener("pointerup", pointerupHandler);
		element.addEventListener("pointerdown", pointerdownHandler);
		element.addEventListener("pointermove", pointermoveHandler);
		element.addEventListener("pointerleave", pointerupHandler);

		return () => {
			element.removeEventListener("pointerup", pointerupHandler);
			element.removeEventListener("pointerdown", pointerdownHandler);
			element.removeEventListener("pointermove", pointermoveHandler);
			element.removeEventListener("pointerleave", pointerupHandler);
		};
	}, [onPanEnd, timelineRef]);
};
```  

You may also need to set the following CSS values on the `root` element for it to behave nicely:  
```css
  user-select: none;
  touch-action: none;
  scroll-behavior: initial;
  -webkit-user-select: none;
  -ms-user-select: none;
  user-select: none;
```
  
Good luck! 🤞

**Reply by arwtyxouymz** [reply]
@samuelarbibe Thank you for the reply!
I was misunderstanding and i could configured both axis scroll based on your solution.

But now, I'm wondering how to stop propagation when dragging item.
It seems `useItem` event handlers don't export DOM event and I can't stop propagation.
https://samuel-arbibe.gitbook.io/dnd-timeline/documentation/item/useitem#events

**Reply by samuelarbibe** [reply]
A few solutions I can recommend: 
1. Make use of [Activation Constraints](https://docs.dndkit.com/api-documentation/sensors/pointer#activation-constraints). The only problem is that they are not implemented into the resizing mechanism, but only into the dragging.
2. Make use of an activation key, like the meta key, and disable the items while pressed.

I see that you have no access to the `activatorEvent` on resize events.  
Let me look into it and understand how I can make it work.

**Reply by samuelarbibe** [reply]
Opened a pull request, where I added an `activatorEvent` to the resize event, that should allow you to call `stopPropagation`.  
I would like you review the PR, and check if it helps you achieve your desired behavior. 
#58 
Thank you!
